### PR TITLE
M8: Enable FlashAttention paths and stabilize training (FA‑2 enablement, autograd‑safe packing, runbook)

### PR DESCRIPTION
**Pull request by seconds-0**
This PR enables FlashAttention in core paths and hardens training stability, along with a focused runbook to validate FA‑2 activation before broader optimizations.

Summary of changes
- FlashAttention dense path:
  - `nsa/kernels/flash_wrappers.py`:
    - `is_flash_available`/`is_flash_varlen_available` now probe real FA‑2 entrypoints.
    - `attention_bgh` calls `flash_attn_func` when available using view/expand reshapes (no `repeat_interleave`), falling back to SDPA on failure.
- FA‑2 varlen (prefill) safety + autograd:
  - `nsa/core/attention_kernels.py`:
    - Sliding: fix `cuq` init; add training‑safe varlen packing via `stack/cat`, preserve workspace path for eval; non‑finite check with masked fallback.
    - Compressed: mirror sliding (training‑safe varlen packing, `cuq/cuk`, finite guard).
- Selection packed training safety:
  - `grouped_selection_attention_packed`: training‑safe bucket packing via `stack/cat` to preserve autograd links; workspace path retained for eval.
- Masked selection NaN guard:
  - `grouped_selection_attention_masked`: zero outputs for rows with no allowed keys (avoids softmax(all −inf)).
- Prefill strict fallbacks (opt‑in):
  - `nsa/core/nsa_attention.py`: under `NSA_STRICT_ASSERTS=1`, detect non‑finite in branch outputs and fallback to safe reference paths.
- FA‑2 gating semantics (in branch already):
  - Hard‑disable via env is honored (`NSA_USE_FA2=0` disables regardless of defaults) using effective flags.
- Docs:
  - Add `Documentation/Runbooks/FlashAttention-Enablement-Test.md` (focused FA‑2 enablement test).
  - Update `CLAUDE.md` with a “FlashAttention Quick Test (GPU)” section linking to the runbook.
  - Add Core Engineer report summarizing sliding varlen OOB fix.

Why
- Prior runs showed flash attention wasn’t actually enabled and hot paths fell back to slow SDPA with `repeat_interleave` materializations.
- Training hit NaNs and in‑place/backward errors due to workspace packing during training.
- These changes enable FA‑2 where available and make training paths autograd‑safe, with defensive fallbacks.

Invariants and constraints
- Causality, GQA consistency, Eq.9 mapping (cmp→sel), Eq.10 group semantics unaffected.
- CPU fallback works; masked SDPA semantics preserved for safe paths.
- Deterministic tests remain deterministic; no Triton added in M0.

How to test (GPU A100)
1) Install FA‑2, enable flags, and run parity tests:
```
uv venv -p 3.11 .venv && source .venv/bin/activate
uv pip sync -r requirements-gpu-cu121-torch24.txt
pip install --no-cache-dir --no-build-isolation flash-attn
NSA_USE_FA2=1 NSA_FA2_MIN_LEN_WIN=1 NSA_FA2_MIN_LEN_CMP=1 \
NSA_TEST_FA2=1 uv run -q pytest -k "fa2_parity or backward_varlen"
```
2) Follow the runbook `Documentation/Runbooks/FlashAttention-Enablement-Test.md` for a short runtime smoke and basic throughput sanity.

Acceptance
- FA‑2 dense/varlen parity tests pass; no non‑finite; no CUDA illegal memory access.
- Short decode/prefill smoke runs cleanly with FA‑2 active; ≥ 300 tok/s on A100 for S≈2048 (informal gate; full production runbook has stricter criteria).

Notes
- Gate MLP init remains per PRD; monitoring of gate stats intact.
- If any branch emits non‑finite values, enable `NSA_STRICT_ASSERTS=1` to force safe fallbacks and capture logs.

### CODE REVIEW

**Comment by seconds-0** [comment]
Thanks for the detailed review — addressed the actionable items:

- Added non‑finite logging:
  - `attention_bgh`: logs `warn.flash_bgh_nonfinite` before `nan_to_num` when FA‑2 dense returns non‑finite.
  - FA‑2 varlen:
    - Sliding: logs `warn.fa2_win_varlen_nonfinite` and `warn.fa2_win_bucket_nonfinite` before masked fallback.
    - Compressed: logs `warn.fa2_cmp_varlen_nonfinite` before masked fallback.
- Env guard improvements:
  - Clamp `NSA_FA2_MIN_LEN_{WIN,CMP}` to minimum 1 after parse.
- Workspace safety:
  - Added bounds assertions in both sliding/compressed varlen pack fill loops (`write_pos + L <= total_k`).
  - Exposed cleanup helpers: `clear_varlen_workspaces()` and `clear_selection_pack_workspaces()` to optionally free global caches.

Leaving as follow‑ups (non‑blocking):
- Thread safety for global workspaces — single‑threaded by design in M0/M1; can wrap in a simple lock if we introduce multithreaded packers later.
- Narrow exception types — we keep broad try/except around FA‑2 calls to preserve robust fallbacks; can tighten if we see noisy failures in practice.
- Numeric env range validation — most are parsed with safe defaults; we’ll extend clamping if new tunables get added.

Let me know if you’d like any of these elevated into tests (e.g., error‑path tests for varlen fallbacks, workspace growth/cleanup). CMT

**Comment by seconds-0** [comment]
Moved the Performance Optimization Plan out of Reports to comply with lint: now lives at Documentation/Plans/M8-Performance-Optimization-Plan.md. PR scope remains focused on FA‑2 enablement and stability; plan doc is preserved for subsequent perf phases.
